File: scripts/stock_codes.py
"""
stock_codes.py — 讀 data/stock_codes.json,提供「代碼 → Yahoo 後綴嘗試順序」。

讓抓 Yahoo 前直接選對後綴(上市 .TW / 上櫃 .TWO),省掉上櫃股先撞 .TW 失敗的無效請求。
未收錄代碼(興櫃/ETF/新掛牌)回預設順序,仍雙試 → 不退化。
對照表由 scripts/build_stock_codes.py 產生。
"""

from __future__ import annotations

import functools
import json
import os

_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
_PATH = os.path.join(_ROOT, "data", "stock_codes.json")


@functools.lru_cache(maxsize=1)
def _codes() -> dict:
    try:
        with open(_PATH, encoding="utf-8") as f:
            return json.load(f).get("codes", {}) or {}
    except (OSError, ValueError):
        return {}


def preferred_suffix(ticker: str) -> str | None:
    s = _codes().get(str(ticker), {}).get("suffix")
    return s if s in (".TW", ".TWO") else None


def suffix_order(ticker: str) -> tuple[str, str]:
    """已知上櫃 → ('.TWO','.TW');其餘(上市/未知)→ ('.TW','.TWO')。一律保留雙試備援。"""
    return (".TWO", ".TW") if preferred_suffix(ticker) == ".TWO" else (".TW", ".TWO")

File: web/src/lib/stockCodes.ts
/**
 * 官方代碼 → Yahoo 後綴對照(上市 .TW / 上櫃 .TWO),來源:TWSE ISIN 官方頁
 * (scripts/build_stock_codes.py → web/src/lib/stockSuffix.json,bundle 進 serverless)。
 *
 * 用途:抓 Yahoo 前先選對後綴,省掉上櫃股「先撞 .TW 失敗再試 .TWO」的無效請求。
 * 未知代碼(興櫃/ETF/新掛牌未收錄)回傳預設順序,仍雙試 → 不會退化。
 */
import suffixMap from "./stockSuffix.json";

const MAP = suffixMap as Record<string, string>;

export type TwSuffix = ".TW" | ".TWO";

/** 官方市場別對應的後綴;未知回 null */
export function preferredSuffix(ticker: string): TwSuffix | null {
  const s = MAP[ticker];
  return s === ".TW" || s === ".TWO" ? s : null;
}

/**
 * 嘗試順序:已知上櫃 → [.TWO, .TW];其餘(上市/未知)→ [.TW, .TWO]。
 * 一律保留兩者作備援,故對未收錄代碼與現行行為完全一致。
 */
export function suffixOrder(ticker: string): readonly [TwSuffix, TwSuffix] {
  return preferredSuffix(ticker) === ".TWO" ? [".TWO", ".TW"] : [".TW", ".TWO"];
}
